rust/kernel/seq_file.rs

```diff
@@ -0,0 +1,221 @@
+    v: *mut c_types::c_void,
+    pos: *mut bindings::loff_t,
+) -> *mut c_types::c_void {
+    if v.is_null() {
+        return ptr::null_mut();
+    }
+
+    // SAFETY: `v` was created by a previous call to `next_callback` or
+    // `start_callback` and both functions return either a null pointer
+    // or pointer generated by `Box<Peekable<T::Iterator>>::into_raw`.
+    let mut iterator = unsafe { Box::from_raw(v as *mut Peekable<T::Iterator>) };
+
+    // SAFETY: The caller guarantees tha `pos` is a valid pointer to an
+    // `loff_t` and expects this function to mutate the value.
+    unsafe {
+        *pos += 1;
+    }
+
+    match iterator.as_mut().next() {
+        Some(_seen) => Box::into_raw(iterator) as *mut c_types::c_void,
+        None => ptr::null_mut(),
+    }
+}
+
+extern "C" fn show_callback<T: SeqOperations>(
+    m: *mut bindings::seq_file,
+    v: *mut c_types::c_void,
+) -> c_types::c_int {
+    // SAFETY: `v` was created by a previous call to `next_callback` or
+    // `start_callback` and both functions return either a null pointer
+    // or pointer generated by `Box<Peekable<T::Iterator>>::into_raw`.
+    let iterator = unsafe { (v as *mut Peekable<T::Iterator>).as_mut() };
+    if let Some(iterator) = iterator {
+        let s = match iterator.peek() {
+            // TODO: Replace with fallible `to_string` when available.
+            Some(item) => item.to_string() + "\0",
+            None => "\0".to_string(),
+        };
+        // SAFETY: Calling a C function. `s` is guaranteed to be null terminated
+        // because we explicitly constructed it just above.
+        unsafe {
+            bindings::seq_puts(m, s.as_ptr() as *const u8 as *const c_types::c_char);
+        }
+    }
+    0
+}
+
+extern "C" fn start_callback<T: SeqOperations>(
+    m: *mut bindings::seq_file,
+    pos: *mut bindings::loff_t,
+) -> *mut c_types::c_void {
+    // SAFETY: This function will be called by opening a proc file generated
+    // by a call to `proc_create::<T>`, so `m` can be converted using
+    // `SeqFile<T>`.
+    let arg = unsafe { &*SeqFile::<T>::convert(m) };
+    // SAFETY: The caller guarantees that `pos` points to a valid `loff_t`.
+    let pos = unsafe { *pos };
+    match T::start(arg) {
+        Some(mut wrapper) => {
+            for _ in 0..pos {
+                wrapper.as_mut().next();
+            }
+            Box::into_raw(wrapper) as *mut c_types::c_void
+        }
+        None => ptr::null_mut(),
+    }
+}
+
+struct SeqFileOperationsVTable<T>(PhantomData<T>);
+
+impl<T: SeqOperations> SeqFileOperationsVTable<T> {
+    const VTABLE: bindings::seq_operations = bindings::seq_operations {
+        start: Some(start_callback::<T>),
+        stop: Some(stop_callback::<T>),
+        next: Some(next_callback::<T>),
+        show: Some(show_callback::<T>),
+    };
+}
+
+/// A `seq_file` referencing data of type `T`.
+pub struct SeqFile<T> {
+    seq_ops: bindings::seq_operations,
```

Comment:
We should be using the `const` version generated above.

Reply:
:+1: